packages/onegraph-apollo-server-auth/src/index.js

```diff
@@ -0,0 +1,159 @@
+// Peers
+const {SchemaDirectiveVisitor} = require('apollo-server');
+const {
+  AuthenticationError,
+  AuthorizationError,
+} = require('apollo-server-express');
+const {GraphQLNonNull} = require('graphql');
+// Deps
+const jwt = require('jsonwebtoken');
+var jwksClient = require('jwks-rsa');
+
+const atob = str => {
+  return Buffer.from(str, 'base64').toString('binary');
+};
+
+class hasRoleDirective extends SchemaDirectiveVisitor {
+  visitFieldDefinition(field) {
+    const requiredRoles = this.args.oneOf;
+    const resolver =
+      field.resolve ||
+      (source => {
+        const value = source[field.name];
+        return value;
+      });
+    const {url} = this.args;
+
+    field.resolve = (source, args, context, info) => {
+      const userRoles =
+        (context.jwt && context.jwt.user && context.jwt.user.roles) || [];
+      const roleRequirementSatisfied = userRoles.some(role =>
+        requiredRoles.includes(role),
+      );
+
+      if (!roleRequirementSatisfied) {
+        throw new AuthenticationError(
+          `Roles required to access ${field.name}: oneOf=[${requiredRoles.join(
+            ', ',
+          )}], roles you have=[${userRoles.join(', ')}]`,
+        );
+      }
+
+      return resolver(source, args, context, info);
+    };
+
+    if (field.type instanceof GraphQLNonNull) {
+      field.type = field.type.ofType;
+    }
+  }
+}
+
+class isAuthenticatedDirective extends SchemaDirectiveVisitor {
+  visitFieldDefinition(field) {
+    const resolver = field.resolve;
+    const {url} = this.args;
+    field.resolve = (source, args, context, info) => {
+      const userId = context.jwt && context.jwt.user && context.jwt.user.id;
```

Comment:
You could use idx here

Reply:
I guess it's just a devDep, so that should be fine right?